src/Common/src/Interop/Windows/Winsock/Interop.WinsockAsync.cs

```diff
@@ -161,14 +165,29 @@ internal struct NetworkEvents
         }
 
         [StructLayout(LayoutKind.Sequential)]
-        internal struct WSAMsg
+        internal unsafe struct WSAMsg
         {
-            internal IntPtr socketAddress;
+            internal byte* socketAddress;
             internal uint addressLength;
-            internal IntPtr buffers;
+            internal WSABuffer* buffers;
             internal uint count;
             internal WSABuffer controlBuffer;
             internal SocketFlags flags;
+            internal int pad;
+
+            // NOTE: Even though Win32 docs say the flags arg is a DWORD, and thus 4 bytes long, it seems to actually write 4 bytes past this.
+            // The "pad" field above deals with this.
+
+            internal unsafe WSAMsg(byte* socketAddressBuffer, int socketAddressLen, WSABuffer* wsaBuffers, int bufferCount, WSABuffer controlBuffer, SocketFlags flags)
```

Comment:
How about having the ctor take uint rather than casting in the ctor?  i.e. push the responsibility to the code initializing this to make sure it's taking underflow possibility into account.